### PR TITLE
feat(envs): Adding RoboCasa to LeRobot

### PR DESCRIPTION
**Pull request by ShahRutav**
## Title

Added RoboCasa environment.

## Type / Scope

- **Type**: (Feature)
- **Scope**: (optional — name of module or package affected)

## Summary / Motivation

- Adds another benchmark to the lerobot pipeline (https://github.com/robocasa/robocasa)

## Related issues

- Fixes / Closes: # (if any)
- Related: #2380 

## What changed

- Short, concrete bullets of the modifications (files/behaviour).
   -  `examples/port_datasets/port_robocasa.py` Ports the RoboCasa dataset to LeRobot format
   -  `examples/training/train_policy_casa.py` Example training script with RoboCasa dataset
   -  `src/lerobot/envs/configs.py` Adds extra RoboCasa config
   - `src/lerobot/envs/factory.py` Adds the intermediate condition to call the RoboCasa env
   - `src/lerobot/envs/robocasa_env.py` RoboCasa env wrapper
- Short note if this introduces breaking changes and migration steps.
   - Vectorized env with `batch_size > 1` leads to some rendering issues

## How was this tested

- Manual checks / dataset runs performed.

To run the code conversion, download the dataset from RoboCasa with images (https://robocasa.ai/docs/use_cases/downloading_datasets.html). Then, port the dataset to Lerobot format using the following:
  ```bash
  python -m robocasa.scripts.download_datasets --ds_types human_im
  python examples/port_datasets/port_robocasa.py --dataset_path /path/to/dataset.hdf5 --repo_name your_hf_username/robocasa_dataset
```

For training a diffusion policy with the generated dataset
```
python examples/training/train_policy_casa.py
```

For running evaluations,
```
lerobot-eval \      
    --policy.path=outputs/train/example_robocasa_diffusion \
    --env.type=robocasa --env.task TurnOnStove \
    --eval.batch_size=1 \
    --eval.n_episodes=20 \
    --policy.use_amp=false \
    --policy.device=cuda
```

- Run a quick example or CLI (if applicable):

  ```bash
  lerobot-train --some.option=true
  ```

## How to run locally (reviewer)

- Run the relevant tests:

## Checklist (required before merge)

- [x] Linting/formatting run (`pre-commit run -a`)
- [ ] All tests pass locally (`pytest`)
- [ ] Documentation updated
- [ ] CI is green

## Reviewer notes

- Anything the reviewer should focus on (performance, edge-cases, specific files) or general notes.
- Anyone in the community is free to review the PR.

## Known Issues 

Evaluation with the vectorized env (batch_size > 1) causes rendering issues.

### CODE REVIEW

**Comment by jadechoghari** [comment]
Great work @ShahRutav ! It would be a huge addition to lerobot, let us know if you need any help :)

**Comment by ShahRutav** [comment]
One remaining gap is that rendering breaks when using synced vectorized environments during evaluation (`batch_size > 1`).
I’ve tracked this down to an upstream `robosuite` and/or `gymnasium` issue rather than anything specific to LeRobot.
See ARISE-Initiative/robosuite#787 for context.

**Comment by jadechoghari** [comment]
That's weird... For instnace libero/metaworld are based on robosuite, and this issue don't happen... Do you have an example for it?
Also, do you have the robosuite lerobot dataset uploded somewhere on the HF Hub?
Thanks!

**Comment by ShahRutav** [comment]
A minimal reproducing example of the error is provided here: https://github.com/ARISE-Initiative/robosuite/issues/787#issuecomment-3704208147

I’ve also uploaded the converted TurnOnStove dataset here: https://huggingface.co/datasets/Rutav/robocasa-TurnOnStove
 (Rutav/robocasa-TurnOnStove).

If you set this dataset name in examples/training/train_policy_casa.py, you should be able to run training with the diffusion policy end-to-end.

**Comment by jadechoghari** [comment]
Thanks ! @ShahRutav do you have a list of all robocasa tasks?

**Comment by ShahRutav** [comment]
Hi @jadechoghari , I haven’t been able to spend time resolving this bug yet. Feel free to take a stab

Here's the list of tasks envs from RoboCasa,
https://github.com/ShahRutav/lerobot/blob/1bfc22ea2c37f074b1675eb6009f869f5b2bdd7e/src/lerobot/envs/robocasa_env.py#L156-L188

**Comment by ShahRutav** [comment]
Hi @jadechoghari,
We’ve identified the root cause of the bug here: https://github.com/ARISE-Initiative/robosuite/issues/787

I tested the fix with the LeRobot policy evaluation (`batch_size > 1`), and the returned images now look correct. I’ll clean up the PR over the next couple of days.
Please feel free to take a look

**Comment by jadechoghari** [comment]
One final thing @ShahRutav , do we have access to all of robocasa dataset tasks?
Thx!

**Comment by ShahRutav** [comment]
It’s publicly available. @jadechoghari  would you prefer that I host the dataset in my HF dataset space, or would you like to host it on your end?

**Comment by jadechoghari** [comment]
We can host it in lerobot org if it's too big, however is it avaialble as a lerobot dataset?

**Comment by ShahRutav** [comment]
The publicly available dataset is not yet in the LeRobot format. This PR draft includes `examples/port_datasets/port_robocasa.py`, which converts the dataset from RoboCasa to LeRobot format.

File: examples/training/train_policy_casa.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    # We can then instantiate the dataset with these delta_timestamps configuration.
+    dataset = LeRobotDataset(
+        task_name, delta_timestamps=delta_timestamps, video_backend="pyav"
+    )  # pyav is slow
+    ##################################################### CHANGE HERE #####################################################
+    # remove unnecessary columns; it will throttle the whole process
+    want = {"observation.state", "action", "timestamp", "index", "episode_index", "task_index", "task"}
+    have = set(dataset.hf_dataset.column_names)
+    drop = [c for c in have if c not in want]
+    dataset.hf_dataset = dataset.hf_dataset.remove_columns(drop)
```
[the remark this answers is not in the document]
Comment (ShahRutav) [reply examples/training/train_policy_casa.py:106]: Is there a better way to do this? @jadechoghari 

The converted dataset has many fields that may not be useful to load every time. Currently, I am deleting those extra fields manually from the dataset. Is there a cleaner way supported?

**Comment by jadechoghari** [comment]
@ShahRutav great work so far! Can we have one big LeRobotDataset of robocasa, we can help with that if you need support!

**Comment by alexcbb** [comment]
Hi, I've been using this PR to train Pi0 with the datasets provided in: https://github.com/huggingface/lerobot/issues/2380#issuecomment-3691091917 One potential problem might be the potential version conflicts between the lerobot library and robocasa for evaluating in the environment. Did you manage to have a proper suite of version for the conflicts (numpy, etc.) ? Seems like this would require to be managed in the pyproject for further proper usage in the lerobot lib

**Comment by ShahRutav** [comment]
> @ShahRutav great work so far! Can we have one big LeRobotDataset of robocasa, we can help with that if you need support!

@jadechoghari, do you want me to upload a single dataset for all the Robocasa datasets, or do you need the script to handle it? I missed your message earlier.


> Hi, I've been using this PR to train Pi0 with the datasets provided in: [#2380 (comment)](https://github.com/huggingface/lerobot/issues/2380#issuecomment-3691091917) One potential problem might be the potential version conflicts between the lerobot library and robocasa for evaluating in the environment. Did you manage to have a proper suite of version for the conflicts (numpy, etc.) ? Seems like this would require to be managed in the pyproject for further proper usage in the lerobot lib

You are right. It has some conflicts in `numpy` as well as `robosuite` versions. I used the following for this PR:
```
numpy==1.23.5
robosuite==1.5.2
```

**Comment by imstevenpmwork** [comment]
Hello everyone, just checking in to see how can I help unblocking this PR

**Comment by jadechoghari** [comment]
Hey guys @ShahRutav @alexcbb do you have any updates on this PR? let us know if you need help

**Comment by ShahRutav** [comment]
@jadechoghari is there anything blocking this PR?

**Comment by jadechoghari** [comment]
@ShahRutav probably training and evaluating on robocasa , have you done this?

**Comment by alexcbb** [comment]
Hi @jadechoghari, I've trained a Pi-0 model a few weeks ago but didn't have time to evaluate yet (got stuck on some conference deadlines). I plan to work a bit on it in the following days as I'm planning to make use of robocasa on my own experiments.

**Comment by ShahRutav** [comment]
> @ShahRutav probably training and evaluating on robocasa , have you done this?

I trained a model on the TurnOnStove task, and the behavior appears qualitatively correct, but the success rate remains zero. Do you have a LeRobot training config @jadechoghari? I suspect the default one in [train_policy.py](https://github.com/huggingface/lerobot/blob/main/examples/training/train_policy.py) isn’t tuned.

**Comment by ShahRutav** [comment]
I have also merged all the atomic tasks in a single dataset here: https://huggingface.co/datasets/Rutav/robocasa_single_stage

**Comment by ShahRutav** [comment]
I have successfully verified the training and evaluation pipelines for two individual tasks: TurnOnSinkFaucet and TurnOnStove. Using the [train_policy.py](https://github.com/huggingface/lerobot/blob/main/examples/training/train_policy.py) example with 5k steps, I achieved the following success rates:

TurnOnSinkFaucet: 15%

TurnOnStove: 20%

@jadechoghari, if you have optimized training configurations or would like to run a parallel evaluation to verify these baselines, I’d welcome the feedback

**Comment by brunomachado37** [comment]
Hello @ShahRutav ,

I wasn't able to execute your code to evaluate a policy. I suspect you are not using the latest version of RoboCasa. Could you give more details on your installation, the exact version (or commit) of each library ?

I would be glad to contribute to the pull request, either by training and evaluating models or by adapting the code to the latest RoboCasa version.

**Comment by ShahRutav** [comment]
Hi @brunomachado37,

I have not tested the code with the latest RoboCasa, released last week. The following installation configuration should work (the commit dates before the RoboCasa365 release):
```
-e git+ssh://git@github.com/robocasa/robocasa.git@756598a5be52e052339bb2d957426e39015c2afb#egg=robocasa
robosuite==1.5.2
mujoco==3.2.6
numpy==1.23.5
``` 

If you can help with testing the latest Robocasa code, that would be amazing!

**Comment by brunomachado37** [comment]
Hello @ShahRutav, unfortunately I could not run your code, probably due to some library version incompatibility that I could not figure it out. Since there are significant changes between the version of RoboCasa you were working on and the latest, me and @alexcbb thought it would be easier to create a new pull request from scratch. Most of the changes here are not compatible or are not needed with the new RoboCasa code.

Here it is the new pull request: https://github.com/huggingface/lerobot/pull/3217

If you could give some feedback on the integration, I would really appreciate it. As far as me and @alexcbb have tested, it is all working smoothly. We took some inspiration on your code, thank you for your collaboration and for the initiative. I hope our integration can be useful for you.

**Comment by ShahRutav** [comment]
Thanks for the help, @brunomachado37 and @alexcbb! Closing this PR, assuming the latest robocasa365 is integrated here: https://github.com/huggingface/lerobot/pull/3217
